Fluidframework documentation for audience example (#12289)
microsoft/FluidFramework: docs/content/docs/recipes/audience-react.md

[IMAGE]

{{< callout note >}}

This tutorial assumes that you are familiar with the [Fluid Framework Overview]({{< relref "/docs/_index.md" >}}) and that you have completed the [QuickStart]({{< relref "quick-start.md" >}}). You should also be familiar with the basics of [React](https://reactjs.org/), [creating React projects](https://reactjs.org/docs/create-a-new-react-app.html#create-react-app), and [React Hooks](https://reactjs.org/docs/hooks-intro.html).

{{< /callout >}}

## Create the project

1. Open a Command Prompt and navigate to the parent folder where you want to create the project; e.g., `c:\My Fluid Projects`.
1. Run the following command at the prompt. (Note that the CLI is np**x**, not npm. It was installed when you installed Node.js.)

    ```dotnetcli
    npx create-react-app fluid-audience-tutorial
    ```

1. The project is created in a subfolder named `fluid-audience-tutorial`. Navigate to it with the command `cd fluid-audience-tutorial`.

1. The project uses the following Fluid libraries:

    |Library |Description |
    |---|---|

>     | `fluid-framework`    |Contains the SharedMap [distributed data structure]({{< relref "dds.md" >}}) that synchronizes data across clients. *This object will hold the most recent timestamp update made by any client.*|

Reviewer: Is there a _timestamp_ in this app?

    | `@fluidframework/azure-client`   |Defines the connection to a Fluid service server and defines the starting schema for the [Fluid container][].|
    | `@fluidframework/test-client-utils`   |Defines the `InsecureTokenProvider` needed to create the connection to a Fluid Service.|
    {.table}

    Run the following command to install the libraries.

    ```dotnetcli
    npm install @fluidframework/azure-client @fluidframework/test-client-utils fluid-framework
    ```

## Code the project

### Set up state variables and component view

> 1. Open the file `\src\App.js` in your code editor. Delete all the default `import` statements. Then delete all the markup from the `return` statement. Then add import statements for components and react hooks. Note that we will be implementing the components in the later steps. The file should look like the following:

Reviewer: "React" is a proper noun and should always be capitalized.
```suggestion
1. Open the file `\src\App.js` in your code editor. Delete all the default `import` statements. Then delete all the markup from the `return` statement. Then add import statements for components and React hooks. Note that we will be implementing the components in the later steps. The file should look like the following:
```

```js
  import { useState, useCallback } from "react";
  import { AudienceDisplay } from "./AudienceDisplay";
  import { UserIdSelection } from "./UserIdSelection"

  export const App = () => {
    // TODO 1: Define state variables to handle view changes and user input
    return (
    // TODO 2: Return view components
    );
  }
```

> 1. Replace `TODO 1` with the following code. Note that the values `userId` and `containerId` will come from a `UserIdSelection` component through the `handleSelectUser` function. The `displayAudience` variable is used to switch between the member list view and the userId selection view. In the case that a user inputs an invalid `audienceId` they will be redirected to the userId selection view by the `handleContainerNotFound` function.

Reviewer 2: Focus the content around why this code is here and not just what is happening. E.g. What does `handleSelectUser` do and why are we setting all these properties?

Avoid referring to concepts that have not been introduced yet. When you do need to do this make sure you tell the reader where they will be introduced. Otherwise people feel like they missed something. E.g. you talk about the member list view and the userId selection view but those are in the next code block.

Use class names (with optional descriptors) instead of just descriptors. Eg. `AudienceDisplay` vs "member list view". 

```suggestion
1. Replace `TODO 1` with the following code. This code initializes local state variables that will be used within the application. `displayAudience` determines if we render the `AudienceDisplay` component or the `UserIdSelection` component (see `TODO 2`). `userId` contains the user identifier to connect to the container with and `containerId` is the container to load. `handleSelectUser` and `handleContainerNotFound` are passed in as callbacks to the two views and manage state transitions. `handleSelectUser` is called when attempting to create/load a container. `handleContainerNotFound` is called when creating/loading a container fails.

Note, the values `userId` and `containerId` will come from a `UserIdSelection` component through the `handleSelectUser` function. 
```

  <SelectUser onSelectUser={handleSelectUser}/>
```

### Set up AudienceDisplay component

1. Create and open a file `\src\AudienceDisplay.js` in your code editor. Add the following `import` statements:

```js
  import { useEffect, useState } from "react";
  import { SharedMap } from "fluid-framework";
  import { AzureClient } from "@fluidframework/azure-client";
  import { InsecureTokenProvider } from "@fluidframework/test-client-utils"
```

1. Add the following functional components and helper functions:

```js
  const tryGetAudienceObject = async (userId, userName, containerId) => {
    // TODO 1: Configure the client and container
    // TODO 2: Configure the container
    // TODO 3: Get the container and services from the Fluid service.
    // TODO 4: Return the Audience object.
  }

  export const AudienceDisplay = (props) => {
    //TODO 5: Configure user id, user name, and state variables
    //TODO 6: Set state variables and set event listener on component mount
    //TODO 7: Return list view
  }

  const AudienceList = (data) => {
    //TODO 8: Append view elements to list array for each member
    //TODO 9: Return list of member elements

>   }

Reviewer: Why do these functions have noun names instead of verb names as functions normally do? And why do they begin with a capital letter? Shouldn't these be named displayAudience and getAudience?

Author: These functions are React function components that return React elements. Since they are alternatives to class components, they tend to follow the same naming conventions as classes.

Reviewer: > These functions are React function components that return React elements. Since they are alternatives to class components, they tend to follow the same naming conventions as classes.

Got it. thanks.

```

### Getting container and audience

> The Fluid runtime will bring changes made to the audience through the container services, but Fluid is agnostic about the UI framework. You can use a helper function to get the Fluid data, from the Audience object into the view layer (the React state). The tryGetAudienceObject method is called when the view component loads after a user id is selected. The returned value is assigned to a React state property.

Reviewer: Not clear what "bring" means here. Bring from what to what?

> The Fluid runtime will bring changes made to the audience through the container services, but Fluid is agnostic about the UI framework. You can use a helper function to get the Fluid data, from the Audience object into the view layer (the React state). The tryGetAudienceObject method is called when the view component loads after a user id is selected. The returned value is assigned to a React state property.

Reviewer: ```suggestion
The Fluid runtime will bring changes made to the audience through the container services, but Fluid is agnostic about the UI framework. You can use a helper function to get the Fluid data, from the Audience object, into the view layer (the React state). The tryGetAudienceObject method is called when the view component loads after a user id is selected. The returned value is assigned to a React state property.
```

> 1. Replace `TODO 1` with the following code. Note that the values for `userId` `userName` `containerId` will be passed in from the `App` component. If there is no `containerId`, a new container is created. Before the client can be used, it needs an `AzureClientProps` that will define the type of connection the client will be using. Think of the `serviceConfig` as the properties required to connect to the service. Note that local mode of Azure Client is used here. To enable collaboration across all client, please replace it with Fluid Relay Service credentials. See how to [set up the Azure Fluid Relay service](https://docs.microsoft.com/en-us/azure/azure-fluid-relay/how-tos/provision-fluid-azure-portal) for more information.

Reviewer: ```suggestion
1. Replace `TODO 1` with the following code. Note that the values for `userId` `userName` `containerId` will be passed in from the `App` component. If there is no `containerId`, a new container is created. Before the client can be used, it needs an `AzureClientProps` that will define the type of connection the client will be using. Think of the `serviceConfig` as the properties required to connect to the service. Note that local mode of Azure Client is used here. To enable collaboration across all clients, please replace it with Fluid Relay Service credentials. See how to [set up the Azure Fluid Relay service](https://docs.microsoft.com/en-us/azure/azure-fluid-relay/how-tos/provision-fluid-azure-portal) for more information.
```

> 1. Replace `TODO 1` with the following code. Note that the values for `userId` `userName` `containerId` will be passed in from the `App` component. If there is no `containerId`, a new container is created. Before the client can be used, it needs an `AzureClientProps` that will define the type of connection the client will be using. Think of the `serviceConfig` as the properties required to connect to the service. Note that local mode of Azure Client is used here. To enable collaboration across all client, please replace it with Fluid Relay Service credentials. See how to [set up the Azure Fluid Relay service](https://docs.microsoft.com/en-us/azure/azure-fluid-relay/how-tos/provision-fluid-azure-portal) for more information.

Reviewer: I don't think the line about FRS should be here. This refers to something a dev would do when they are moving their own app from dev to production. It doesn't belong here in a tutorial where all clients are on the same computer.

      list.push(
        <div style={{
          padding: '1rem',
          margin: '1rem',
          display: 'flex',
          outline: 'solid',
          flexDirection: 'column',
          maxWidth: '25%',
          outlineColor
        }} key={key}>
          <div style={{fontWeight: 'bold'}}>Name</div>
          <div>
              {member.userName}
          </div>
          <div style={{fontWeight: 'bold'}}>ID</div>
          <div>
              {member.userId}
          </div>
          <div style={{fontWeight: 'bold'}}>Connections</div>
          {
              member.connections.map((data, key) => {
                  return (<div key={key}>{data.id}</div>);
              })
          }
          <div style={{fontWeight: 'bold'}}>Additional Details</div>
          { JSON.stringify(member.additionalDetails, null, '\t') }
        </div>
      );
  });
```

1. Replace `TODO 9` with the following code.

```js
  return (
      <div>
          {list}
      </div>
  );
```

### Setup SelectUser component

1. Create and open a file `\src\SelectUser.js` in your code editor. Add the following `import` statements and functional components:

```js
import { useState } from 'react';

> export const SelectUser = (props) => {

Reviewer: should this be named selectUser (camel case) as JS functions normally are?

Author: This is also a function component so it should probably be renamed to a noun.

Reviewer: Got it. Thanks.

  // TODO 1: Define styles and handle user inputs
  return (
  // TODO 2: Return view components
  );
}
```

1. Replace `TODO 1` with the following code. Note that the `onSelectUser` function will update the state variables in the parent `App` component and prompt a view change.

```js

>   const selectStyle = {

Reviewer: An object with a verb name? Should this be selectionStyle?

  <UserIdSelection onSelectUser={handleSelectUser}/>
```

### Set up AudienceDisplay component

1. Create and open a file `\src\AudienceDisplay.js` in the code editor. Add the following `import` statements:

```js
  import { useEffect, useState } from "react";
  import { SharedMap } from "fluid-framework";
  import { AzureClient } from "@fluidframework/azure-client";
  import { InsecureTokenProvider } from "@fluidframework/test-client-utils";
```

Note that the objects imported from the Fluid Framework library are required for defining users and containers. In the following steps, `AzureClient` and `InsecureTokenProvider` will be used to configure the client service (see `TODO 1`) while the `SharedMap` will be used to configure a `containerSchema` needed to create a container (see `TODO 2`).

1. Add the following functional components and helper functions:

```js
  const tryGetAudienceObject = async (userId, userName, containerId) => {
    // TODO 1: Configure the client and container
    // TODO 2: Configure the container
    // TODO 3: Get the container and services from the Fluid service.

>     // TODO 4: Return the Audience object.

Reviewer 2: I want to lean into your earlier point of not trying to teach everything. Let's just group this all together as 1 TODO with the description of reviewing https://fluidframework.com/docs/recipes/react/ and https://fluidframework.com/docs/build/containers/ if the developer want's to learn more about container creation.

```js
  useEffect(() => {
    tryGetAudienceObject(userId, userName, props.containerId).then(audience => {
      if(!audience) {
        props.onContainerNotFound();
        alert("error: container id not found.");
        return;
      }

      const updateMembers = () => {
        setFluidMembers(audience.getMembers());
        setCurrentMember(audience.getMyself());
      }

      updateMembers();

      audience.on("membersChanged", updateMembers);

      return () => { audience.off("membersChanged", updateMembers) };
    });
  }, []);
```

1. Replace `TODO 7` with the following code. Note, if the `fluidMembers` or `currentMember` has not been initialized, a blank screen is rendered. The `AudienceList` component will render the member data with styling (to be implemented in the next section).

```js
  if (!fluidMembers || !currentMember) return (<div/>);

  return (

>       <div>
>         <AudienceList fluidMembers={fluidMembers} currentMember={currentMember}/>
>       </div>

Reviewer 2: These divs are redundant.

```suggestion
        <AudienceList fluidMembers={fluidMembers} currentMember={currentMember}/>
```
